### ISSUE TITLE
Cells are rerendered on every click

### ISSUE THREAD

**Issue by rgavrilov** [mui/mui-x#7807]
### Order ID 💳

52426

### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### The problem in depth 🔍

Same as:
https://github.com/mui/mui-x/issues/3480#issue-1084636084

I just need you to confirm that a trivial UI operation is still causing unnecessary rerendering of all the cells (which is ridiculous).

Sandbox:
https://codesandbox.io/s/datagrid-rerendering-issue-tzdri2

just click around and watch Value column being rerendered.

The issue:
In my actual use case cells are rendered with 4 buttons inside. Even scrolling becomes too sluggish. I'm questioning whether this component would be usable in the UI. 

### Your environment 🌎

<details>
  <summary>`npx @mui/envinfo`</summary>

```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```
</details>

**Comment by rgavrilov** [comment]
At the very least, I need an example of how to cache a rendered cell. I have tried caching a result of `renderCell`:

```
            renderCell: (params: GridRenderCellParams<Response, Row, any>) => {
                const response = params.value!;

                if (cache[params.rowNode.id]) {
                    return cache[params.rowNode.id];
                } else {
                    console.log("missed cache");
                    return cache[params.rowNode.id] = <CheckboardCell response={ response }/>;
                }

            },
```

but that didn't help much. Even assuming that response doesn't change, the `CheckboardCell` is being rerendered.

**Comment by yaredtsy** [comment]
I have checked it with the profiler and everything is being rendered for some reason.

https://user-images.githubusercontent.com/34584266/216557645-1a8ff067-5967-4442-b0a2-f5347160b8c0.mp4

I think it should have been like this.

https://user-images.githubusercontent.com/34584266/216560002-d370c078-4c90-4885-9606-d381a13096d1.mp4

**Comment by yaredtsy** [comment]
hey @m4theushw, anything you do in the datagrid causes every componet(ColumnHeader, footer, row, cell) to rerender even moving the mouse in and out of the datagrid is causing rerender.

https://user-images.githubusercontent.com/34584266/216837874-ee54a703-300e-499f-81da-1d6ff9d4cb09.mp4

**Comment by yaredtsy** [comment]
it's perfect now, the only problem I had was, with the custom cell, I had to memorize it myself.

**Comment by rompeldunk** [comment]
@m4theushw : Really appreciation if memorize of rows, cells and column can get in place!
The model should be to create custom Components like you do in the PR link?

Will test your commit and report back: https://pkg.csb.dev/mui/mui-x/commit/794239b6/@mui/x-data-grid <3

**Comment by rompeldunk** [comment]
Would it be possible to test an equivalent PR to x-data-grid-pro?
I tried mixing the commit with my existing code and memoize it, but it it broke, as expected.

We're trying to render many hundreds, up to many thousand of rows. One checkboxSelection event renders the whole stack. Also tested with renderCell disabled.

```js
<DataGridPro
  rows={rows}
  columns={browseColumns}
  columnBuffer={3}
  rowHeight={32}
  autoHeight
  autoPageSize
  components={{
    // Memoized comp's here
  }}
  disableSelectionOnClick={true}
  checkboxSelection
  rowBuffer={3}
  onSelectionModelChange={handleSetSelectedRowIds}
  onCellEditCommit={handleCellEditCommit}
  onColumnHeaderClick={handleColumnHeaderClick}
  />

```

**Comment by cherniavskii** [comment]
@rompeldunk

> Would it be possible to test an equivalent PR to x-data-grid-pro?

Yes, you can find links to the packages in the Packages section of Codesandbox CI build: https://ci.codesandbox.io/status/mui/mui-x/pr/7846/builds/344182

**Comment by rompeldunk** [comment]
> @rompeldunk
> 
> > Would it be possible to test an equivalent PR to x-data-grid-pro?
> 
> Yes, you can find links to the packages in the Packages section of Codesandbox CI build: https://ci.codesandbox.io/status/mui/mui-x/pr/7846/builds/344182

I notice that there's a lot of breaking changes. Maybe I've picked version 6 etc?

I tested this one:
https://pkg.csb.dev/mui/mui-x/commit/794239b6/@mui/x-data-grid-pro

Currently using "@mui/x-data-grid-pro": "5.15.1",

**Comment by rompeldunk** [comment]
Just to clarify: (in my use-case at least)

- It is the params from the `renderCell (params: GridRenderCellParams)` which cause the problems, since i.e. the `api` reference which changes for every render?
- Wouldn't using a callback instead (to a `state, setState`) with none-changing ref instead do the trick?

- Would an alternative optimization-technique be to use the `components `and `componentsProps` props?

**Comment by 0x2539** [comment]
I'm still seeing unnecessary re-renders on scroll/click/edit, but it's just for the visible rows/cells (not the whole table). It makes no difference if I have a custom `renderCell` or not. Is this expected @m4theushw?

The scrolling performance is good, my only problem is that when I click in a cell to start editing there is a slight delay until the cell turns to a textfield. My impression is that at worst it should re-render the row (all the cells in the row) of the cell that you are editing, and at best, it would re-render just the clicked cell

**Comment by LinnaViljami** [comment]
Still facing the problem. The whole DataGrid is rerendered on every cell or row click, all rows and cells are rendered. It causes a significant performance issue with a table over about hundred cells
